web/packages/teleport/src/DesktopSession/DesktopSession.tsx

```diff
@@ -58,92 +51,97 @@ declare global {
   } = props;
 
-  const processing =
-    fetchAttempt.status === 'processing' ||
-    tdpConnection.status === 'processing';
+  const [state, setState] = useState<{
+    screen: 'processing' | 'anotherSessionActive' | 'canvas' | 'alert dialog';
+    alertMessage?: string;
+  }>({
+    screen: 'processing',
+  });
 
-  // onDialogClose is called when a user
-  // dismisses a non-fatal error dialog.
-  const onDialogClose = () => {
-    // The following state-setting calls will
-    // cause the useEffect below to calculate the
-    // errorDialog state.
+  useEffect(() => {
+    setState(prevState => {
+      // We always want to show the user the first alert that caused the session to fail/end,
+      // so if we're already showing an alert, don't change the screen.
+      //
+      // This allows us to track the various pieces of the state independently and always display
+      // the vital information to the user. For example, we can track the TDP connection status
+      // and the websocket connection status separately throughout the codebase. If the TDP connection
+      // fails, and then the websocket closes, we want to show the TDP connection error to the user,
+      // not the websocket closing. But if the websocket closes unexpectedly before a TDP message telling
+      // us why, we want to show the websocket closing message to the user.
+      if (prevState.screen === 'alert dialog') {
+        return prevState;
+      } else {
+        // Otherwise, calculate screen state:
+        const showAnotherSessionActive = showAnotherSessionActiveDialog;
+        const showAlert =
+          fetchAttempt.status === 'failed' || // Fetch attempt failed
+          tdpConnection.status === 'failed' || // TDP connection failed
+          tdpConnection.status === '' || // TDP connection ended gracefully server-side
+          wsConnection.status === 'closed'; // Websocket closed (could mean client side graceful close or unexpected close, the message will tell us which)
+        const processing =
+          fetchAttempt.status === 'processing' ||
+          tdpConnection.status === 'processing';
 
-    setTdpConnection(prevState => {
-      if (prevState.status === '') {
-        // If prevState.status was a non-fatal error,
-        // we assume that the TDP connection remains open.
-        return { status: 'success' };
+        if (showAnotherSessionActive) {
+          return { screen: 'anotherSessionActive' };
+        } else if (showAlert) {
+          let message = '';
+          if (fetchAttempt.status === 'failed') {
+            message = fetchAttempt.statusText || 'fetch attempt failed';
+          } else if (tdpConnection.status === 'failed') {
+            message = tdpConnection.statusText || 'TDP connection failed';
+          } else if (tdpConnection.status === '') {
+            message =
+              tdpConnection.statusText || 'TDP connection ended gracefully';
+          } else if (wsConnection.status === 'closed') {
```

Comment:
The number of nested if/else blocks in here is 🤯

Any way we can simplify this?

Reply:
I did another minor overhaul to make the state machine more comprehensible: 6cfb564322a1deedff79e1451b2d5982d8f92bbe